### PR TITLE
Purge all cache with callback

### PR DESCRIPTION
**Pull request by waahm7**
*Issue #, if available:*

*Description of changes:*
- Adds a purge_cache_with_callback function which has a callback.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/host_resolver.c [hunk 1]
```diff
@@ -311,25 +320,88 @@ static void s_remove_host_listener_from_entry(
...
     for (struct aws_hash_iter iter = aws_hash_iter_begin(table); !aws_hash_iter_done(&iter);
          aws_hash_iter_next(&iter)) {
         struct host_entry *entry = iter.element.value;
+        if (on_purge_cache_complete_callback) {
+            /* acquire a refernce to wait for the callback to trigger */
+            aws_ref_count_acquire(&purge_callback_options->ref_count);
+            entry->on_host_purge_complete = s_purge_cache_callback;
+            entry->on_host_purge_complete_user_data = purge_callback_options;
+        }
         s_shutdown_host_entry(entry);
```
Comment (TingDaoK) [inline source/host_resolver.c:379]: I think this is not worth to be a separate call. Always freak me out when we held the lock and call into other functions. And this function is just too simple and only invoked once.

File: include/aws/io/host_resolver.h [hunk 2]
```diff
@@ -238,10 +241,19 @@ AWS_IO_API int aws_host_resolver_record_connection_failure(
     struct aws_host_address *address);
 
 /**
+ * @Deprecated Use aws_host_resolver_purge_cache_v2 instead
  * calls purge_cache on the vtable.
  */
 AWS_IO_API int aws_host_resolver_purge_cache(struct aws_host_resolver *resolver);
 
+/**
+ * calls purge_resolver_cache on the vtable.
```
[the remark this answers is not in the document]
Comment (waahm7) [reply include/aws/io/host_resolver.h]: Thanks, fixed.

File: include/aws/io/host_resolver.h [hunk 3]
```diff
@@ -238,10 +241,19 @@ AWS_IO_API int aws_host_resolver_record_connection_failure(
...
+ */
+AWS_IO_API int aws_host_resolver_purge_cache_v2(
```
[the remark this answers is not in the document]
Comment (waahm7) [reply include/aws/io/host_resolver.h]: Thanks, I have renamed it to `aws_host_resolver_purge_cache_with_callback`

File: source/host_resolver.c [hunk 4]
```diff
@@ -311,25 +320,88 @@ static void s_remove_host_listener_from_entry(
...
     }
 
+    /* release the original reference */
+    if (on_purge_cache_complete_callback) {
+        aws_ref_count_release(&purge_callback_options->ref_count);
```
[the remark this answers is not in the document]
Comment (waahm7) [reply source/host_resolver.c]: Thanks, that is a great catch. I have refactored the overall logic as trying to reuse `s_clear_default_resolver_entry_table_synced` was making the overall code complicated.

File: source/host_resolver.c (hunk 1 above)
Comment (waahm7) [reply source/host_resolver.c]: Thanks, fixed.

File: source/host_resolver.c [hunk 5]
```diff
@@ -325,11 +366,78 @@ static void s_clear_default_resolver_entry_table(struct default_host_resolver *r
     aws_hash_table_clear(table);
 }
 
-/* TODO: add a callback to notify when purge has completed */
 static int resolver_purge_cache(struct aws_host_resolver *resolver) {
```
Comment (TingDaoK) [inline source/host_resolver.c:369]: rename this to `s_resolver_purge_cache`

File: source/host_resolver.c [hunk 6]
```diff
@@ -325,11 +366,78 @@ static void s_clear_default_resolver_entry_table(struct default_host_resolver *r
...
+    if (!on_purge_cache_complete_callback) {
+        return resolver_purge_cache(resolver);
+    }
+
+    struct default_host_resolver *default_host_resolver = resolver->impl;
+    aws_mutex_lock(&default_host_resolver->resolver_lock);
+    struct aws_hash_table *table = &default_host_resolver->host_entry_table;
+
+    /* Success if hash table is already empty. */
+    if (!aws_hash_table_get_entry_count(table)) {
```
Comment (TingDaoK) [inline source/host_resolver.c:416]: super trivial: for clarity
```suggestion
    if (aws_hash_table_get_entry_count(table) == 0) {
```

File: source/host_resolver.c [hunk 7]
```diff
@@ -325,11 +366,78 @@ static void s_clear_default_resolver_entry_table(struct default_host_resolver *r
...
+        aws_mutex_lock(&entry->entry_lock);
+        entry->on_host_purge_complete = s_purge_cache_callback;
+        entry->on_host_purge_complete_user_data = purge_callback_options;
+        entry->state = DRS_SHUTTING_DOWN;
+        aws_mutex_unlock(&entry->entry_lock);
+    }
+    aws_hash_table_clear(table);
+
+    /* release the original reference */
+    aws_ref_count_release(&purge_callback_options->ref_count);
```
Comment (TingDaoK) [inline source/host_resolver.c:440]: move this outside the lock. the entry will finish purging from another thread, still possible for this release to be the last refcount.

File: source/host_resolver.c [hunk 8]
```diff
@@ -325,11 +366,78 @@ static void s_clear_default_resolver_entry_table(struct default_host_resolver *r
...
+    struct default_host_resolver *default_host_resolver,
+    aws_simple_completion_callback *on_purge_cache_complete_callback,
+    void *user_data) {
+
+    struct host_purge_callback_options *purge_callback_options = s_host_purge_callback_options_new(
+        default_host_resolver->allocator, on_purge_cache_complete_callback, user_data);
+    purge_callback_options->allocator = default_host_resolver->allocator;
+
+    struct aws_task *task = aws_mem_calloc(default_host_resolver->allocator, 1, sizeof(struct aws_task));
+    aws_task_init(task, s_purge_host_cache_callback_task, purge_callback_options, "async_purge_host_address_task");
```
Comment (TingDaoK) [inline source/host_resolver.c:395]: The task is not doing the purge work, it only exists to invoke callback asynchronously.
```suggestion
    aws_task_init(task, s_purge_host_cache_callback_task, purge_callback_options, "async_purge_host_callback_task");
```

**Review by TingDaoK — APPROVED** [review]
Fix&ship

File: source/host_resolver.c (hunk 5 above)
Comment (waahm7) [reply source/host_resolver.c]: Sorry, old code. I have fixed it.

File: source/host_resolver.c (hunk 8 above)
Comment (waahm7) [reply source/host_resolver.c]: Thanks, fixed.

**Pull request merged**